### PR TITLE
Embed safe/vault targets into repos

### PR DESCRIPTION
**Pull request by dennisjbell**
Similar to the reliance on external ~/.bosh/config for targeting bosh, we rely on the the user's safe config (.saferc and .svtoken) to target the vault. While this almost works for single user, it breaks down when spread across a team, and can cause significant damage when a user forgets that they're targeting a different vault when switching context.

Instead, Genesis will track the vault used to create the deployment internally, and target it prior to any actions requiring access to the vault

See https://trello.com/c/n4WhOC6p in Trello for full details

### CODE REVIEW

**Comment by dennisjbell** [comment]
There are several "supportive" commits in here that were not explicitly part of this fix, but were written generically so that they could be used beyond how they were utilized for this particular refactor of the Vault addressing system.

File: lib/Genesis.pm [hunk 1]
```diff
@@ -785,6 +793,15 @@ This is best used with C<run()>, like this:
 
     my @lines = lines(run('some command'));
 
+==head2 read_json_from($out, $rc)
+
+Ignore C<$rc>, and parses C<$out> as JSON, returning the resulting structure.
```
[the remark this answers is not in the document]
Comment (jhunt) [reply lib/Genesis.pm]: Also of note, there's an extra `=` on the `=head2`

File: lib/Genesis/Env.pm [hunk 2]
```diff
@@ -732,18 +732,14 @@ sub _check_secret {
...
 		}
 
 	} else {
-		if (@missing) {
-			explain("  #R{\x{2718}}  %s [#C{%s}]", $path, $type);
-			explain("     #R{\x{2718}}  :%s", $_) for (@missing);
-		} else {
-			explain("  #G{\x{2714}}  %s [#C{%s}]", $path,$type);
-		}
+		bullet(@missing ? "bad" : "good", sprintf("%s [#C{%s}]", $path, $type));
```
[the remark this answers is not in the document]
Comment (jhunt) [reply lib/Genesis/Env.pm]: "forwards compatible"?

File: lib/Genesis/Env.pm [hunk 3]
```diff
@@ -735,12 +735,13 @@ sub purge_secrets {
 
 	die_unless_controlling_terminal($self);
 
-	explain "#Ry{[WARNING]} The following existing secrets will be removed:";
+	explain "#Yr{[WARNING]} The following pre-existing secrets will need to be removed:";
 	bullet $_ for (@paths);
 	my $response = prompt_for_line(undef, "Type 'yes' to remove these secrets","");
 	if ($response eq 'yes') {
-		explain "\nDeleting existing secrets under '#C{%s}'...", $self->prefix;
+		{local $/ = ''; explain "\nDeleting existing secrets under '#C{%s}'...", $self->prefix;}
```
[the remark this answers is not in the document]
Comment (jhunt) [reply lib/Genesis/Env.pm]: I'm actually confused right now why we are setting the line separator for an explain...

[the remark this answers is not in the document]
Comment (dennisjbell) [reply lib/Genesis/Env.pm]: was hacking around the fact that explain always printed a newline, and I didn't want a new line.  Resolving it in a different way...

File: lib/Genesis/Kit.pm [hunk 4]
```diff
@@ -179,6 +179,8 @@ sub run_hook {
 		$ENV{GENESIS_ROOT}         = $opts{env}->path;
 		$ENV{GENESIS_ENVIRONMENT}  = $opts{env}->name;
 		$ENV{GENESIS_TYPE}         = $opts{env}->type;
+		$ENV{GENESIS_TARGET_VAULT} = $opts{env}->vault->url;
+		$ENV{GENESIS_VERIFY_VAULT} = $opts{env}->vault->verify || "";
```
Comment (dennisjbell) [inline lib/Genesis/Kit.pm]: Set 'SAFE_TARGET' to target vault, so that `safe -T` isn't needed

**Comment by dmolik** [comment]
👍

**Pull request merged**